src/components/input/input.ts

```diff
@@ -0,0 +1,168 @@
+/**
+ * @license
+ *
+ * Copyright IBM Corp. 2019
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { customElement, LitElement, html, property } from 'lit-element';
+import classnames from 'classnames';
+import settings from 'carbon-components/es/globals/js/settings';
+import WarningFilled16 from '@carbon/icons/lib/warning--filled/16';
+import styles from './input.scss';
+
+const { prefix } = settings;
+
+/**
+ * Supported input types.
+ *
+ * For this component we only support textual types
+ */
+export enum INPUT_TYPE {
+  EMAIL = 'email',
+  PASSWORD = 'password',
+  TEL = 'tel',
+  TEXT = 'text',
+  URL = 'url',
+}
+
+/**
+ * Input element. Supports all the usual attributes for textual input types
+ */
+@customElement(`${prefix}-input`)
+export default class BXInput extends LitElement {
+  /**
+   * May be any of the standard HTML autocomplete options
+   */
+  @property()
+  autocomplete = '';
+
+  @property({ type: Boolean })
+  autofocus = false;
+
+  @property({ type: Boolean, reflect: true })
+  disabled = false;
+
+  /**
+   * The helper text. Corresponds to `helper-text` attribute.
+   */
+  @property({ attribute: 'helper-text' })
+  helperText = '';
+
+  @property()
+  id = '';
+
+  @property({ type: Boolean, reflect: true })
+  invalid = false;
```

Comment:
Discussion; Reflecting the discussion we had in React codebase, this codebase stopped supporting "empty validity message" scenario and thus `--invalid` class is added based on `validityMessage` being non-empty (e.g. https://github.com/carbon-design-system/carbon-custom-elements/blob/29ad0c9/src/components/dropdown/dropdown.ts#L481), instead of using another property. Wondering if this make sense to you or you think separate `invalid` is more desirable despite the discussion of "empty validity message". What do you think?

Reply:
Since the invalid message can be an element `slot`ted in to the component, it seems to make more sense to keep the simpler `invalid` toggle instead of requiring a full DOM insertion.

We should be able to check for the existence of a `slot`ted element, or a valid `validityMessage` and only then apply the invalid state - that way we can avoid the whole "empty validity message" issue (if that's the design we're set on)

Reply:
Good point on `@slotchange`! Though I couldn't rattle off a record that led to in favor of reduced properties (attributes), this is what I recall (and thus this codebase took such approach for now). Another bonus here is that introducing `invalid` attribute/prop (back) won't be a breaking change (though the opposite is a breaking change and thus React codebase still has `invalid` prop).

Reply:
Update: We are keeping both `invalid` and `validity-message`, at least for now, given there is not much connection in the meaning of those attributes.